scripts/receive-payjoin.py

```diff
@@ -4,19 +4,80 @@
-    open_test_wallet_maybe, get_wallet_path, check_regtest, \
-    add_base_options
+from twisted.web.server import Site
+from twisted.internet import reactor
+import txtorcon
+from jmbase import get_log, set_logging_level, jmprint
+from jmclient import jm_single, load_program_config, \
+    WalletService, open_test_wallet_maybe, get_wallet_path, check_regtest, \
+    add_base_options, PayjoinServer
 from jmbase.support import EXIT_FAILURE, EXIT_ARGERROR
-from jmbitcoin import amount_to_sat
+from jmbitcoin import amount_to_sat, amount_to_btc, encode_bip21_uri, \
+     CCoinAddress
+
+# txtorcon outputs erroneous warnings about hiddenservice directory strings,
+# annoyingly, so we suppress it here:
+import warnings
+warnings.filterwarnings("ignore")
 
 jlog = get_log()
 
-def receive_payjoin_main(makerclass):
+bip78_amount = None
+bip78_receiving_address = None
+
+def setup_failed(arg):
+    jlog.error("SETUP FAILED", arg)
+    reactor.stop()
+
+def create_onion_ep(t, hs_public_port):
+    return t.create_onion_endpoint(hs_public_port)
+
+def onion_listen(onion_ep, site):
+    return onion_ep.listen(site)
+
+def print_host(ep):
+    """ Callback fired once the HS is available;
+    receiver user needs a BIP21 URI to pass to
+    the sender:
+    """
+    jmprint("Your hidden service is available. Please now pass "
+            "this URI string to the sender to effect the payjoin "
+            "payment:")
+    jmprint(bip21_uri_from_onion_hostname(str(ep.getHost().onion_uri),
+                                          ep.getHost().onion_port))
+    jmprint("Keep this process running until the payment is received.")
+
+def bip21_uri_from_onion_hostname(host, port):
+    """ Encoding the BIP21 URI according to BIP78 specifications,
+    and specifically only supporting a hidden service endpoint.
+    Note: we hardcode http; no support for TLS over HS.
+    Second, note we convert the global amount-in-sats bip78_amount
+    to BTC denomination as expected by BIP21.
+    """
+    full_pj_string = "http://" + host + ":" + str(port)
+    bip78_btc_amount = amount_to_btc(amount_to_sat(bip78_amount))
+    # "safe" option is required to encode url in url unmolested:
+    return encode_bip21_uri(bip78_receiving_address,
+                            {"amount": bip78_btc_amount,
+                             "pj": full_pj_string.encode("utf-8")},
+                            safe=":/")
+
+def start_tor(site, hs_public_port):
+    d = txtorcon.connect(reactor)
+    d.addCallback(create_onion_ep, hs_public_port)
+    d.addErrback(setup_failed)
+    d.addCallback(onion_listen, site)
+    d.addCallback(print_host)
+
+def receive_payjoin_main():
+    global bip78_receiving_address
+    global bip78_amount
+
     parser = OptionParser(usage='usage: %prog [options] [wallet file] [amount-to-receive]')
     add_base_options(parser)
+    parser.add_option('-P', '--hs-port', action='store', type='int',
+                      dest='hsport', default=7081,
```

Comment:
Why port 7081 as default, not 80?

Reply:
Good question. Do you think 80 is a good idea? I kind of like explicitly choosing a non-standard port but perhaps that's a bad idea.

Reply:
I don't see a downside of using standard port. It's actually better, as we can remove ":80" part from URL and make it three bytes shorter. :)

Reply:
Addressed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/682/commits/06d8772b366fba84f63e252c772c465219376b9d